sklearn/impute.py

```diff
@@ -624,12 +642,15 @@ def _impute_one_feature(self,
         # get posterior samples
         X_test = safe_indexing(X_filled[:, neighbor_feat_idx],
                                missing_row_mask)
-        mus, sigmas = predictor.predict(X_test, return_std=True)
-        good_sigmas = sigmas > 0
-        imputed_values = np.zeros(mus.shape, dtype=X_filled.dtype)
-        imputed_values[~good_sigmas] = mus[~good_sigmas]
-        imputed_values[good_sigmas] = self.random_state_.normal(
-            loc=mus[good_sigmas], scale=sigmas[good_sigmas])
+        if self.sample_posterior:
```

Comment:
I would make a check that the signature of `predictor.predict` has a `return_std` to raise a proper error.

Reply:
@jnothman suggested it's not necessary above. The only way I found to do this is:

```
predictor = BayesianRidge()
'return_std' in list(inspect.signature(predictor.predict).parameters.keys()) # True

predictor = RandomForestRegressor()
'return_std' in list(inspect.signature(predictor.predict).parameters.keys()) # False
```

Reply:
Uhm that's ugly.

In the same time, it is true that the error is ok
```
In [8]: reg.predict(X, predict_std=True)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-d45a86188634> in <module>()
----> 1 reg.predict(X, predict_std=True)

TypeError: predict() got an unexpected keyword argument 'predict_std'
```

Reply:
Agreed on both points. Let's leave as is then.

Reply:
Fwiw you don't need list or keys in the snippet above